File: 2003/IntroToVBNETPart2/Presentation/ParameterfulProperties.cs
using System ;
using System.Collections ;

public class DependentCollection
{
    private ArrayList collection ;
    private Hashtable lookup ;

    public DependentCollection ()
    {
        collection = new ArrayList() ;
        lookup = new Hashtable() ;
    }

    public virtual Dependent this [ int index ]
    {
        get { return (Dependent) collection [ index ] ; }
        set 
        { 
            collection [ index ] = value ;
            lookup.Add ( value.FirstName, value ) ;
        }
    }
    public virtual Dependent this [ string firstName ]
    {
        get { return (Dependent) lookup [ firstName ] ; }
        set 
        { 
            Dependent dependent = (Dependent) lookup [ firstName ] ;
            if ( dependent == null )
                Add ( value ) ;
            else
            {
                for ( int i = 0 ; i < collection.Count ; i++ )
                {
                    dependent = (Dependent) collection [ i ] ;
                    if ( dependent.FirstName == value.FirstName )
                    {
                        collection.RemoveAt ( i ) ;
                        Add ( value ) ;
                    }
                }
            }
        }
    }

    public void Add ( Dependent value )
    {
        collection.Add ( value ) ;
        lookup.Add ( value.FirstName, value ) ;
    }

    public void Delete ( Int32 Index )
    {
        Dependent dependent ;
        dependent = (Dependent) collection [ Index ] ;
        if ( dependent != null ) 
        {
            lookup.Remove ( dependent.FirstName ) ;
            collection.RemoveAt ( Index ) ;
        }
    }

    public void Delete ( string firstName )
    {
        Dependent dependent ;
        int i ;
        for ( i = 0 ; i < collection.Count ; i++ )
        {
            dependent = (Dependent) collection[i] ;
            if ( dependent.FirstName == firstName )
            {
                lookup.Remove ( firstName ) ;
                collection.RemoveAt ( i ) ;
            }
        }
    }
    public int Count 
    {
        get { return collection.Count ; }
    }
}

public class Test
{
    public static void Main ()
    {
        DependentCollection dc = new DependentCollection () ;
        dc.Add ( new Dependent ( "Nathan", new DateTime ( 1986, 1, 20 )  ) ) ;
        dc.Add ( new Dependent ( "Daryl", new DateTime ( 1988, 5, 29 ) ) ) ; 
        dc.Add ( new Dependent ( "Joshua", new DateTime ( 1992, 2, 17 ) ) ) ; 
        for ( int i = 0 ; i <= dc.Count - 1 ; i++ )
            Console.WriteLine ( dc [ i ] ) ;
        Dependent d ;
        d = dc [ "Nathan" ] ;
        Console.WriteLine ( "Nathan: {0}", d ) ;
        dc.Delete ( "Nathan" ) ;
        Console.WriteLine ( "After deleting dependent Nathan" ) ;
        for ( int i = 0 ; i <= dc.Count - 1 ; i++ )
            Console.WriteLine ( dc [ i ] ) ;
    }
}

File: 2003/IntroToVBNETPart2/Presentation/ParameterfulProperties.vb
Option Strict On

Imports System
Imports System.Collections

Public Class DependentCollection
    Private collection As ArrayList
    Private lookup As Hashtable

    Public Sub New()
        collection = New ArrayList()
        lookup = New Hashtable()
    End Sub

    Default Public Overridable Property Item(ByVal index As Int32) As Dependent
        Get
            Return DirectCast(collection.Item(index), Dependent)
        End Get
        Set(ByVal Value As Dependent)
            collection.Item(index) = Value
            lookup.Add(Value.FirstName, Value)
        End Set
    End Property

    Default Public Overridable Property Item(ByVal firstName As String) As Dependent
        Get
            Return DirectCast(lookup.Item(firstName), Dependent)
        End Get
        Set(ByVal Value As Dependent)
            Dim dependent As Dependent = DirectCast(lookup.Item(firstName), Dependent)
            If dependent Is Nothing Then
                Add(Value)
            Else
                Dim i As Integer
                For i = 0 To collection.Count
                    dependent = DirectCast ( collection.Item(i), Dependent )
                    If dependent.FirstName = Value.FirstName Then
                        collection.RemoveAt(i)
                        Add(Value)
                    End If
                Next
            End If
        End Set
    End Property

    Public Sub Add(ByVal Value As Dependent)
        collection.Add(Value)
        lookup.Add(Value.FirstName, Value)
    End Sub

    Public ReadOnly Property Count() As Int32
        Get
            Return collection.Count()
        End Get
    End Property

    Public Sub Delete(ByVal Index As Int32)
        Dim dependent As Dependent
        dependent = DirectCast ( collection.Item(Index), Dependent )
        If Not dependent Is Nothing Then
            lookup.Remove(dependent.FirstName)
            collection.RemoveAt(Index)
        End If
    End Sub

    Public Sub Delete(ByVal firstName As String)
        Dim dependent As Dependent
        Dim i As Integer
        For i = 0 To collection.Count
            dependent = DirectCast ( collection.Item(i), Dependent )
            If dependent.FirstName = firstName Then
                lookup.Remove(firstName)
                collection.RemoveAt(i)
            End If
        Next
    End Sub
End Class

Public Class Test
    Shared Sub Main ()
        Dim dc As New DependentCollection ()
        dc.Add ( new Dependent ( "Nathan", #1/20/1986# ) )
        dc.Add ( new Dependent ( "Daryl", #5/29/1988# ) )
        dc.Add ( new Dependent ( "Joshua", #2/17/1992# ) )
        Dim i As Integer
        For i = 0 To dc.Count () - 1
             Console.WriteLine ( dc.Item ( i ) )
        Next i
        Dim d As Dependent
        d = dc.Item ( "Nathan" )
        Console.WriteLine ( "Nathan: {0}", d )
    End Sub
End Class
